### PR TITLE
RavenDB-15369 - UnhandledException during server startup

### PR DESCRIPTION
**Pull request by grisha-kotler**
https://issues.hibernatingrhinos.com/issue/RavenDB-15369

### CODE REVIEW

File: src/Raven.Server/Utils/AffinityHelper.cs [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+using Sparrow.Collections;
+using Sparrow.Logging;
+using Sparrow.Platform;
+using Sparrow.Server.Platform.Posix;
+
+namespace Raven.Server.Utils
+{
+    public class AffinityHelper
+    {
+        private static readonly ReaderWriterLockSlim _affinityLocker = new ReaderWriterLockSlim();
```
Comment (ppekrol) [inline src/Raven.Server/Utils/AffinityHelper.cs]: You can use: EasyReaderWriterLock

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Utils/AffinityHelper.cs [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+                }
+            }
+            else
+            {
+                bitMask = processAffinityMask.Value;
+            }
+
+            _affinityLocker.EnterWriteLock();
+
+            using (_affinityLocker.EnterWriteLock())
```
Comment (ayende) [inline src/Raven.Server/Utils/AffinityHelper.cs]: Why do you have two of these?

File: src/Raven.Server/Utils/AffinityHelper.cs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+            {
+                process.ProcessorAffinity = new IntPtr(bitMask);
+
+                // changing the process affinity resets the thread affinity
+                // we need to change the custom affinity threads as well
+                foreach (var pooledThread in _customAffinityThreads)
```
Comment (ayende) [inline src/Raven.Server/Utils/AffinityHelper.cs]: I'm pretty sure that once we set it at the process level, it applies to the threads too

File: src/Raven.Server/Utils/AffinityHelper.cs [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+                if (result != 0)
+                    throw new InvalidOperationException(
+                        $"Failed to set affinity for thread: {pooledThread.CurrentUnmanagedThreadId}, " +
+                        $"affinity: {affinity}, result: {result}, error: {Marshal.GetLastWin32Error()}");
+            }
+        }
+
+        internal static void RemoveCustomAffinityThread(PoolOfThreads.PooledThread pooledThread)
+        {
+            using (_affinityLocker.EnterReadLock())
```
Comment (ayende) [inline src/Raven.Server/Utils/AffinityHelper.cs]: I really don't like the locking that we have here.

File: src/Raven.Server/Utils/AffinityHelper.cs (hunk 1 above)
Comment (grisha-kotler) [reply src/Raven.Server/Utils/AffinityHelper.cs]: done

**Review by ayende — CHANGES_REQUESTED** [review]
I don't like this approach.
It would be better to kill the locks entirely and simply retry from the current process state on failure a few time, then give up

File: src/Raven.Server/Utils/AffinityHelper.cs (hunk 3 above)
Comment (grisha-kotler) [reply src/Raven.Server/Utils/AffinityHelper.cs]: It does. But it's changing it to the Process affinity and we need it to be 1 less (by default).

File: src/Raven.Server/Utils/AffinityHelper.cs (hunk 4 above)
Comment (ppekrol) [reply src/Raven.Server/Utils/AffinityHelper.cs]: What is the problem of using RWLS?

Comment (grisha-kotler) [reply src/Raven.Server/Utils/AffinityHelper.cs]: moved it to `ResetThreadAffinity`

File: src/Raven.Server/Utils/AffinityHelper.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Raven.Server/Utils/AffinityHelper.cs]: leftover from changed to `EasyReaderWriterLock`. removed.

File: src/Raven.Server/Utils/AffinityHelper.cs (hunk 3 above)
Comment (ayende) [reply src/Raven.Server/Utils/AffinityHelper.cs]: Okay

File: src/Raven.Server/Utils/AffinityHelper.cs (hunk 4 above)
Comment (ayende) [reply src/Raven.Server/Utils/AffinityHelper.cs]: There is no need to do _any_ locking. That is handled by the OS
We just need to handle retries.
Affinity changes can also come from outside.

Comment (grisha-kotler) [reply src/Raven.Server/Utils/AffinityHelper.cs]: done

File: src/Raven.Server/Utils/AffinityHelper.cs [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+            // we need to change the custom affinity threads as well
+            foreach (var pooledThread in _customAffinityThreads)
+            {
+                SetCustomThreadAffinity(pooledThread);
+            }
+        }
+
+        internal static bool ResetThreadAffinity(PoolOfThreads.PooledThread pooledThread)
+        {
+            _customAffinityThreads.TryRemove(pooledThread);
```
Comment (ayende) [inline src/Raven.Server/Utils/AffinityHelper.cs:68]: What about removing this when the thread shuts down? Can we add an explicit handle for that when the thread exit?

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Server/Utils/AffinityHelper.cs (hunk 5 above)
Comment (grisha-kotler) [reply src/Raven.Server/Utils/AffinityHelper.cs:68]: done
https://github.com/ravendb/ravendb/pull/10853/commits/2039600eff1c54add86773d92b75d83a06d84c4b

**Review by ayende — APPROVED** [review]

**Pull request merged**